### PR TITLE
some fix related to vcontest

### PR DESCRIPTION
**Pull request by crccw**

### CODE REVIEW

File: js/contest.js [hunk 1]
```diff
@@ -6,6 +6,13 @@ function escapeHtml(unsafe) {
       .replace(/"/g, "&quot;")
       .replace(/'/g, "&#039;");
 }
+function getDuration($obj){
```
Comment (51isoft) [inline js/contest.js]: put it inside ready()

File: js/contest.js [hunk 2]
```diff
@@ -6,6 +6,13 @@ function escapeHtml(unsafe) {
...
+        var hms=$obj.val().replace(/[,;.-\/]/g,":").replace(/[^0-9:]/g,"").split(":");
```
Comment (51isoft) [inline js/contest.js]: I suspect that there will be a datetime related function, if not, find a plugin to do that :)

File: js/contest.js [hunk 3]
```diff
@@ -75,6 +82,44 @@ $(document).ready(function() {
...
+        end_time.setTime(start_time+duration*1000);
+        if(end_time.toString()!="Invalid Date") $("input[name=end_time]").val($.format.date(end_time.toString(),"yyyy-MM-dd HH:mm:ss"));
+
+        var lock_time=new Date();
+        lock_time.setTime(end_time.getTime()-lockbefore);
+        if(lock_time.toString()!="Invalid Date") $("input[name=lock_board_time]").val($.format.date(lock_time.toString(),"yyyy-MM-dd HH:mm:ss"));
+    }
+    $("input[name=duration]").change(function() {
+        var duration=getDuration($(this));
+        var $text=$(this).parent().parent().parent().next().find("td");
```
Comment (51isoft) [inline js/contest.js]: Add a class, then use parent with selector. This is way too complicated.

File: js/contest.js [hunk 4]
```diff
@@ -75,6 +82,44 @@ $(document).ready(function() {
         window.location.href="contest.php?virtual=1";
     });
 
+    function updateTimeSelect(start_time,duration){
+        var lockbefore=Date.parse($("input[name=end_time]").val())-Date.parse($("input[name=lock_board_time]").val());
+        var end_time=new Date();
+        end_time.setTime(start_time+duration*1000);
+        if(end_time.toString()!="Invalid Date") $("input[name=end_time]").val($.format.date(end_time.toString(),"yyyy-MM-dd HH:mm:ss"));
```
Comment (51isoft) [inline js/contest.js]: I don't like this condition, I prefer to use parseInt(start_time) to check whether that's valid.

File: js/contest.js [hunk 5]
```diff
@@ -75,6 +82,44 @@ $(document).ready(function() {
...
+        var $text=$(this).parent().parent().parent().next().find("td");
+        var hh=Math.round(duration/3600);
+        var mm=Math.round(duration%3600/60);
+        var ss=duration%60;
+        if(duration<30*60 || duration>15*24*3600){
+            $text.addClass("text-warning");
+        }else{
+            $text.removeClass("text-warning");
+        }
+        $(this).val(hh+":"+(mm<10?"0":"")+mm+":"+(ss<10?"0":"")+ss);
```
Comment (51isoft) [inline js/contest.js:89]: find a plugin for that :)

File: js/contest.js [hunk 6]
```diff
@@ -112,7 +157,7 @@ $(document).ready(function() {
 
     $(".ptype").change(function() {
         var ptp=$(this).val();
-    //    alert(ptp);
+        //    alert(ptp);
```
Comment (51isoft) [inline js/contest.js]: remove

File: js/contest_edit.js [hunk 7]
```diff
@@ -1,3 +1,10 @@
+function getDuration($obj){
```
Comment (51isoft) [inline js/contest_edit.js]: ditto, consider move that to a common js file

**Comment by 51isoft** [comment]
If you have time, please really consider building v4. I can help dev that on the weekend as well :)

File: js/contest.js (hunk 4 above)
Comment (crccw) [inline js/contest.js]: that won't solve the problem. It's for the case where duration is tooooooooo long.

File: js/contest.js (hunk 3 above)
Comment (crccw) [inline js/contest.js]: I have add id to the prompting text.. although i think rewrite that form without the use of table is the best way..

File: js/contest_edit.js (hunk 7 above)
Comment (crccw) [inline js/contest_edit.js]: I see no need for it to be in global.js, since it's only used when creating/editing contests

**Comment by 51isoft** [comment]
ooops, rebase this?

**Comment by crccw** [comment]
I'll rewrite this part.

**Pull request closed without merging**